Split CLI and client

bigspider opened the pull request:
Changes in this PR:
- Added an RPC server, moved the get_all_appointments method from the API to the RPC
- Added a cli to interact with the RPC server, with its own config in ~/.teos_cli/
- Added synchronization using a ReadWriteLock, since the API server and RPC server are on two separate threads
- <strike>Replaced the Flask development server with gevent.pywsgi.WSGIServer to serve both the API and the RPC Flask apps</strike> - reverted
- Moved the client (the former cli) to the new /contrib folder, with settings in ~/.teos_client/
- Refactored TowerResponseError and TowerConnectionError into common

Still in progress:
- two unit tests for the get_all_appointments method that was moved to the RPC are currently moved to the e2e and commented out.

A larger refactoring of the cli and the rpc (together with proper tests) will be continued as part of #182.

Closes: #146

sr-gi commented on teos/cli/teos_cli.py:
```diff
@@ -0,0 +1,131 @@
+from common.tools import setup_data_folder
+from common.exceptions import InvalidKey, InvalidParameter, SignatureError, TowerResponseError
+
+from . import DEFAULT_CONF, DATA_DIR, CONF_FILE_NAME
```
Same as for the client

bigspider replied:
Changed to absolute imports in 01bc372.

sr-gi commented on test/teos/e2e/teos.conf:
```diff
@@ -11,6 +11,10 @@ btc_feed_connect = localhost
 btc_feed_port = 28332
 
 [teos]
+api_bind = localhost
```
This are the exact same as the defaults, so I don't think they are actually necessary.  In fact, some of the other may also be taken off.

bigspider replied:
Removed api_bind, api_port, rpc_bind, rpc_port, and all the btc_feed_* settings.

sr-gi commented on test/teos/e2e/test_basic_e2e.py:
```diff
@@ -525,3 +534,79 @@ def test_appointment_shutdown_teos_trigger_while_offline():
     teosd_process.terminate()
+
+
+# def test_get_all_appointments_watcher(api, client, get_all_db_manager):
```
Why are this two commented out?

bigspider replied:
get_all_appointments was moved from the client to the cli, so I thought I'd just figure out how to do proper cli and rpc tests in the next PR together with all the other cli commands.
It is used anyway in the e2e test, so it is not completely not covered.

bigspider commented on .circleci/config.yml:
```diff
@@ -44,10 +44,10 @@ jobs:
           command: |
             . venv/bin/activate
-            pytest test/cli/unit
+            pytest contrib/**/test
```
Yes, that matches all /test folders in any subfolder of contrib/.

bigspider commented on .circleci/config.yml:
```diff
@@ -20,7 +20,7 @@ jobs:
             pip install -r requirements.txt
             pip install -r requirements-dev.txt
-            pip install -r cli/requirements.txt
+            pip install -r contrib/client/requirements.txt
```
Right, added in 7e61a8a4426ef42f34ba395373de310cc963569b.

bigspider commented on contrib/client/teos_client.py:
```diff
@@ -11,21 +13,20 @@
 from common.tools import is_256b_hex_str, is_locator, compute_locator, is_compressed_pk
 
-logger = get_logger()
+from . import DEFAULT_CONF, DATA_DIR, CONF_FILE_NAME
```
Tbh I think relative imports do have an advantage here: as I expect each subfolder of /contrib to be an independent small "package", it has the advantage that it would still work even if you rename the folder.
Also, while we still want to be able to import `contrib.<stuff>` as it's useful for e2e tests, I feel like it should not be the practice, as /contrib stuff is not meant to be part of the "real" codebase. Basically, I wouldn't treat /contrib as a module throughout the codebase.

sr-gi replied:
Seems fair for the contrib bit, since as you said it'll be a standalone part only meant for developing / testing.

This does not apply for the `cli` though, where the same approach was taken.

sr-gi replied:
Just tested this as is and it does not work for me with the relative imports. Have you tested it?

bigspider commented on contrib/client/README.md:
```diff
@@ -1,6 +1,6 @@
+# teos_client
 
-`teos_cli` is a command line interface to interact with the Eye of Satoshi watchtower server, written in Python3.
+`teos_client` is a command line interface to interact with the Eye of Satoshi watchtower server, written in Python3.
```
Added a more extended description in a091bba.

bigspider commented on contrib/client/__init__.py:
```diff
@@ -1,13 +1,13 @@
-    "LOG_FILE": {"value": "teos_cli.log", "type": str, "path": True},
+    "LOG_FILE": {"value": "teos_client.log", "type": str, "path": True},
     "APPOINTMENTS_FOLDER_NAME": {"value": "appointment_receipts", "type": str, "path": True},
     "CLI_PRIVATE_KEY": {"value": "cli_sk.der", "type": str, "path": True},
```
Renamed in a091bba (also cli_sk.der ==> user_sk.der).

bigspider commented on contrib/client/teos_client.py:
```diff
@@ -1,3 +1,5 @@
+#!/usr/bin/env python3
```
Added in a091bba.

bigspider commented on teos/teosd.py:
```diff
@@ -59,6 +62,13 @@ def get_config(command_line_conf, data_dir):
     return config
 
 
+def rpc_thread_function(config, rw_lock, inspector, watcher):
```
Well, I need a function to pass as `target` when I create the thread. Perhaps I could use an inner function, but a separate function makes it explicit what the thread is being given access to (as no variable is automatically brought into scope).

sr-gi replied:
I think it'll be the same as having a `start` or `run`  function in the RPC in the same way we have it in the API.

bigspider commented on contrib/client/INSTALL.md:
```diff
@@ -0,0 +1,38 @@
+# Install 
```
Created a separate issue #186.

bigspider commented on teos/api.py:
```diff
@@ -66,6 +61,8 @@ class API:
     Args:
         host (:obj:`str`): the hostname to listen on.
         port (:obj:`int`): the port of the webserver.
+        rw_lock (:obj:`RWLockWrite <readwritelock.rwlock.RWLockWrite>`): lock that must be acquired before reading or
```
Cool. Tbh, I assumed I should use the same type of lock as the API - but I now had a read of https://en.wikipedia.org/wiki/Readers%e2%80%93writers_problem to understand it better :+1:

bigspider commented on contrib/client/help.py:
```diff
@@ -1,17 +1,16 @@
     return (
         "USAGE: "
-        "\n\tpython teos_cli.py [global options] command [command options] [arguments]"
+        "\n\tpython teos_client.py [global options] command [command options] [arguments]"
```
As this is also dependent on installing teos properly (e.g. by creating symlinks), I'd leave this as part of #186 if you agree.

sr-gi replied:
Agreed

bigspider commented on teos/api.py:
```diff
@@ -74,24 +71,26 @@ class API:
-        self.app = app
+
+        # ToDo: #5-add-async-to-api
+        self.app = Flask(__name__)
```
Added in 2a1405a.

bigspider commented on teos/teosd.py:
```diff
@@ -3,14 +3,17 @@
 from getopt import getopt, GetoptError
 from signal import signal, SIGINT, SIGQUIT, SIGTERM
+import threading
 
```
Done, and moved the `import daemon` after all the standard library imports. 2a1405a

bigspider commented:
Removed most of the redundant docs for the cli (except README), removed the requirements.txt and sharing now data folder and config with teosd.
Removed the logger for the cli as well.

sr-gi commented on teos/api.py:
```diff
@@ -66,32 +61,37 @@ class API:
-        self.app = app
+
+        # ToDo: #5-add-async-to-api
+        self.app = Flask(__name__)
```
The app may need to be outside the object so it can be called by `gunicorn` or `uWSGI` (not completely sure about this yet).

Leaving this here until I can figure it out.

sr-gi commented on teos/cli/teos_cli.py:
```diff
@@ -0,0 +1,131 @@
+#!/usr/bin/env python3
```
I'm not gonna get to picky since the cli needs a followup anyway, but I think the responses should be at least properly formatted for now. Currently this is returning the whole JSON-RPC response, whereas it should only return the "result" field. e.g:

```
teos/cli/teos_cli.py get_all_appointments

{
    "id": 96225753724115027851564137027831193896,
    "jsonrpc": "2.0",
    "result": {
        "responder_trackers": {},
        "watcher_appointments": {}
    }
}

bigspider replied:
Had already done this in the other branch indeed; ported here as well with cee0d72.
